hkmc2/jvm/src/test/scala/hkmc2/JSBackendDiffMaker.scala

```diff
@@ -108,6 +108,10 @@ abstract class JSBackendDiffMaker extends MLsDiffMaker:
               // Otherwise, it is considered a simple runtime error.
               raise(ErrorReport(msg"${message}" -> N :: Nil,
                 source = Diagnostic.Source.Runtime))
+            if traceJS.isSet then
+              output("All outputs:")
```

Comment:
What's with this weird "All outputs" thing? Why would the logging output preceding an error only appear when `traceJS.isSet`? I don't understand you you make this sort of decision. Is there a relevant rationale, from the point of view of the user of the system?

Reply:
PS: Implementation or historical details are not relevant rationales.

Reply:
I guess you were not thinking about the fact that the output is useful even when no tracing is ongoing. Indeed, that's the output of printing stuff, which is currently being silenced when errors occurs (very annoying).

Reply:
I previously chose to only show all outputs when `traceJS` is set because I found out that in most of the tests, "all output"s are just the duplicates of the error messages:
![image](https://github.com/user-attachments/assets/ef8fb4f3-0600-4142-ba6b-d996563c75d9)

But indeed I now see that the output is useful even when no tracing is ongoing, I think maybe I can just not print out the lines showing duplicated messages and "undefined"?

Reply:
> Or maybe I can another flag which controls whether to print all output when error?

No, please! Not another flag. Again, I ask you to think about the _semantics_ of what we're implementing here. How could adding a flag like this that controls low-level details in surprising ways be any good?

Rather, you should figure out why the runtime error appears part of the retrieved output. I believe errors are actually guarded by some invisible unicode characters. Just check how this part of the code works and simply _fix it_. It's not very complicated at all, as far as I remember.

And of course we should also not print the final `undefined`. I'm confident you can figure it out.

Reply:
Thanks, I have just pushed the changes to filter error messages and final `undefined` in the retrieved output for `ReplHost.Error`. Following the implementation in `ReplHost.scala` I realized that uncaught errors are not guarded by special invisible characters, so I just filter them using the index of substring "Uncaught " ([here](https://github.com/CrescentonC/mlscript/blob/4f4083bd017ee9d3d924db80c6888f4236d30b73/hkmc2/jvm/src/test/scala/hkmc2/ReplHost.scala#L56) and [here](https://github.com/CrescentonC/mlscript/blob/4f4083bd017ee9d3d924db80c6888f4236d30b73/hkmc2/jvm/src/test/scala/hkmc2/ReplHost.scala#L51)); And others are filtered out [by using the invisible character](https://github.com/CrescentonC/mlscript/blob/4f4083bd017ee9d3d924db80c6888f4236d30b73/hkmc2/jvm/src/test/scala/hkmc2/ReplHost.scala#L88)

Reply:
Hmmm... But what if someone prints a message that includes the string "Uncaught "?

Reply:
Yeah then the test will fail due to a spurious error caused by the message starting with "Uncaught" that is printed... The special invisible characters for caught errors are added by us in [`ReplHost.scala`](https://github.com/hkust-taco/mlscript/blob/76b129a2bb2daac94ddca9bb6180b654d0044841/hkmc2/jvm/src/test/scala/hkmc2/ReplHost.scala#L107) instead of by nodejs, so the detection of uncaught errors currently are entirely dependent on [this two special strings](https://github.com/hkust-taco/mlscript/blob/76b129a2bb2daac94ddca9bb6180b654d0044841/hkmc2/jvm/src/test/scala/hkmc2/ReplHost.scala#L142-L143)

<img width="344" alt="image" src="https://github.com/user-attachments/assets/6cfc87fb-8cf1-4364-9b08-041cdce5c014">

Reply:
Well that's nothing short of revolting, but I guess we don't have a better way for now...